[FEAT] Event 상세 조회 시 조회수 증가 로직 추가

mkshie opened the pull request:
## 개요
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
<!---- Resolves: #(Issue Number) -->

행사 상세 조회 시 **조회수 집계(총합·일별)**와 **최근 조회 기록(USER/GUEST)** 을 남기고,  
회원/비회원 조회 시나리오를 통합 테스트로 검증하는 기능을 추가했습니다.  
기존 경고(@Builder) 및 EventAction 업데이트 로직도 함께 정리했습니다.

## 🔍 주요 변경 내용

### 1. 일별 조회수 집계(EventViewDaily)

- `EventViewDaily` 엔티티에 `viewDate(LocalDate)` 필드 추가
  -  **viewDate 말고 CreateAt 로 하려고 했는데 그럼 불필요한 코드가 너무 많이 생겨서 viewDate 를 추가하는게 더 좋다고 생각했습니다!**
- `(event_id, view_date)` 유니크 제약으로 **하루에 이벤트당 한 row**만 유지
- `EventViewDailyRepository`
  - `upsertToday(eventId)`  
    - `INSERT ... ON DUPLICATE KEY UPDATE cnt = cnt + 1` 로 일별 조회수 증가
  - `findByEventAndViewDate(Event event, LocalDate viewDate)`  

### 2. 조회수 증가 로직
- `EventGetDetail` 시 `ReadEvent` 추가 
  - `EventViewService.recordView(Long eventId)` 추가
    - `eventViewDailyRepository.upsertToday(eventId)` 호출로 **일별 조회수 +1**
    - `eventRepository.incrementViews(eventId)` 호출로 **이벤트 총 조회수 +1**
  - `EventAction` 생성 및 업데이트

### 3. 최근 조회 기록(EventAction)
- `BaseEntity`
  - `setUpdatedAt()` 헬퍼 메서드 추가 (수정 시점 강제 갱신용)
 - `EventService.getEventDetail(...)`
  - USER / GUEST 조회 시 `ReadEvent(...)` 호출
  - 기존 `EventAction`가 존재하면 `updatedAt`만 갱신
  - 없으면 새로운 `EventAction` 생성

## ✅ 테스트

- `EventServiceTest`
  - `getEventDetail_User_Success`
    - **일반 유저**가 행사 상세 조회 시:
      - `Event.viewsCount` 가 1 증가하는지 검증
      - `EventViewDaily`에 오늘 날짜 기준 `cnt = 1` row 생성 확인
      - `EventAction`에 `actorType = USER`, `actionType = VIEW` 기록 생성 확인
    - 이후 **비회원(guest_id)** 조회 시:
      - 조회수/일별 조회수가 2로 증가하는지 검증
      - GUEST에 대한 `EventAction` 기록 생성 및 `actorType = GUEST` 확인

## PR 유형
어떤 변경 사항이 있나요?
- [x] 새로운 기능 추가
- [x] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.  Commit message convention 참고  (Ctrl + 클릭하세요.)
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트)

mkshie committed:
feat : 행사 조회수·최근 조회 기록 추가

- EventViewDaily viewDate 추가
- EventService ReadEvent 추가
 - EventViewService 의 recordView 를 통해 전체 조회수 증가 및 viewDaily 업데이트
 - eventAction 업데이트 또는 추가
- EventViewDailyRepository 에 event 와 날짜로 찾는 함수 추가

mkshie committed:
feat : 행사 조회수·최근 조회 기록 추가

- BaseEntity 의 updateAt 수정해줄 함수 추가
 - 최근본 행사 조회 시 createAt 순이아닌 updateAt 순으로 변경
- event 조회 시 EventAction 에 동일한 행이 있다면 updateAt 만 수정하게끔

mkshie committed:
refactor(event) : builder.default 추가

mkshie committed:
refactor(event_action) : setter 추가

mkshie committed:
test(EventService) : 코드 정렬 및 조회 회원/비회원 조회 시나리오 추가

- 일반 유저 , 비회원 유저 조회 시 조회수 증가 확인 로직 test 추가

kimjunghwannn commented on src/main/java/com/example/skillup/domain/event/entity/EventAction.java:
EventAction 세터를 열어둔 이유가 있을까요??
EventAction이 생성 되고 난 이후에는 변경될 일이 없는 엔티티기도 해서 세터는 닫아놔도 괜찮을 거 같습니다!

kimjunghwannn commented on src/main/java/com/example/skillup/domain/event/entity/EventViewDaily.java:
BaseEntity의 createdAt이나 updatedAt 대신 준우님이 만드신 viewDate를 기준으로 체크하는 게 더 나아보이네요!
준우님 생각은 어떨지 궁금합니다
날짜 변경도 존재하다 보니 viewDate를 수정해주면서 쓰는 게 좋아보이는데 준우님 생각은 어떠신가요?? 지금 수정하신 아래 쿼리도 createdAt이나 updatedAt 두 가지를 다 보고 비교를 해야하는데 그것보다는 viewDate만 보고 판단하는게 좋아 보입니다!
    List<Event> findRecentEventsByActorId(@Param("actorId")String actorId, Pageable pageable)

mkshie replied:
음... EventAction 에 viewDate 를 사용하자는건가요?? 
findRecentEventsByActorId 해당함수의 쿼리문은 createdAt 이나 updatedAt 를 안보는것처럼 보이는데..!

제가 이해를 못해서 다시 설명해 주실 수 있을까요?

kimjunghwannn replied:
```

       SELECT ea.event
        FROM EventAction ea
        WHERE ea.actorId = :actorId
        ORDER BY ea.updatedAt DESC""")
    List<Event> findRecentEventsByActorId(@Param("actorId")String actorId, Pageable pageable);
```
    
 요 부분 말씀드린 겁니다! 
findRecentEventsByActorId()에서는 현재 EventAction의 updatedAt 기준으로 최근 이벤트를 조회하고 있습니다.

하지만 EventAction 엔티티는 같은 이벤트와 같은 유저일 경우 새로운 엔티티를 만들지 않고 엔티티의 updatedAt만 수정합니다. 이에 지금 저희의 시스템에서 문제가 발생하는데

```
       Optional<EventAction> eventAction = eventActionRepository.findByEventAndActorIdAndActionType(event, actorId,ActionType.VIEW);

        if (eventAction.isPresent()) {
            eventAction.get().setUpdatedAt();
            return;
        }
```


1. updatedAt이 null인 경우가 존재함
예를 들어 11월 26일에 생성된 EventAction이 한 번도 수정되지 않았다면
updatedAt = null이고,
대신 11월 25일에 생성된 EventAction이 25일에 수정되었다면
updatedAt = 11월 25일이 됩니다.

→ 이 경우 26일 데이터가 오히려 더 오래된(=null) 값으로 간주되어
정렬 시 25일 데이터보다 뒤로 밀리는 문제가 발생합니다.

2. 최근 3개월 이내 조회 같은 로직을 createdAt 기준으로 필터링 문제
사용자가 오래된 게시글을 다시 보았더라도
createdAt은 바뀌지 않기 때문에
“최근 본 기록”으로 판단되지 않습니다.

✨ 결론 (핵심)

이러한 문제 때문에 createdAt과 updatedAt 중 어떤 값을 기준으로 삼아야 할지 매번 판단해야 하고,
케이스마다 서로 다른 기준을 쓰게 되어 로직이 복잡해집니다.

그래서 EventAction에 viewDate(또는 actionDate) 같은
“오직 하나의 기준 시간”을 두는 것이 좋을 거 같습니다!

mkshie replied:
아 이해했습니다. 하지만 제 생각엔 
1번의 경우 생성할때 createAt 랑 updateAt 랑 동일한값으로 생성되어서 updateAt 가 null 인경우가 따로 없을거예요.

2번의 경우에 최근 3개월 이내 조회 로직의 경우에는 ea.createAt -> updateAt 로 변경했기때문에 createAt 를 사용하지 않고 조회를 하고 있습니다.

사실 event_view_daily 에 viewDate 를 추가한 이유가 cnt 를 올리는 과정에서 LocalDateTime 을 두가지를 따로 만들어서 그 안에 있는 createAt 와 비교해야하는 과정과 존재하는지 여부에서도 동일한 과정을 반복해야하는 작업때문에 만들었거든요! 

그리고 나서 EventAction 을 봤는데 이 로직이 사용되는 부분이 updateAt 만으로도 view 와 apply 의 경우에는  event 와 user 의 값만으로도 특정되기때문에 updateAt 대신 다른 값이 필요가 없다고 생각했습니다!

kimjunghwannn replied:
아 맞네요 그러면 그대로 가도 괜찮을 거 같습니다!

mkshie replied:
코드리뷰 감사합니다!

kimjunghwannn reviewed (commented):
고생하셨습니다

mkshie committed:
refactor : updateAt 바뀌지 않는 로직 수정 , SampleDataLoader 에서 view_Date 값 삽입

- 네이티브 쿼리로 값을 변경하기 때문에 updateAt 가 조회시에 변경되지 않는 오류 때문에 수정했습니다.
- SampleDataLoader 에서 test 용 회원에 맞게 test 용 user 를 생성하게 수정
- view_date 값 삽입 로직 추가
